feat: add TestLessThanOrEqual hint

LazarusAA opened the pull request:
Hi @zmalatrax

**Changes made:**

- Implemented and tested the TestLessThanOrEqual hint

- Here are the test results:

![image_2024-08-08_060055453](https://github.com/user-attachments/assets/c2cca1d7-c6dd-49fa-bcdb-ad0405d24347)

Please let me know if any additional modifications are needed, and I appreciate any possible feedback.

zmalatrax commented on src/hints/math/testLessThanOrEqual.ts:
```diff
@@ -0,0 +1,55 @@
+ * TestLessThanOrEqual hint
+ *
+ * Check whether the value at `lhs` is less than or equal to the value at `rhs`
+ *
+ * Store the boolean result (0 or 1) at `dst`
+ */
+export const testLessThanOrEqual = (
```
use a proper JSDoc as in [`AllocateFelt252Dict`](https://github.com/kkrt-labs/cairo-vm-ts/blob/e133561a28ea1f269f6efd1af8bb657007c20b74/src/hints/dict/allocFelt252Dict.ts#L28-L42)

zmalatrax reviewed (commented):
great :)

LazarusAA commented:
Hi @zmalatrax

Fixed the documentation and the program comparison. I changed the PR to ready for review but if there's any modification needed please let me know.

zmalatrax commented on src/hints/math/testLessThanOrEqual.ts:
```diff
@@ -0,0 +1,62 @@
+/**
+ * TestLessThanOrEqual hint
+ *
+ * Store true at `dst` if value at `lhs` is inferior or equal to value at `rhs`.
+ * Store false otherwise
```
Better used for `testLessThanOrEqual()` than the type, you can switch it with

Small correction: "Store true at `dst` if value at `lhs` is inferior or equal to value at `rhs`, false otherwise."

"Check whether the value at `lhs` is less than or equal to the value at `rhs`"

LazarusAA replied:
I've made the changes, but I wasn't entirely sure If I understood you correctly. Could you please review it to ensure everything is fine?

zmalatrax commented on src/hints/math/testLessThanOrEqual.ts:
```diff
@@ -0,0 +1,62 @@
+ *
+ * Check whether the value at `lhs` is less than or equal to the value at `rhs`
+ *
+ * Store the boolean result (0 or 1) at `dst`
```
```suggestion
```

zmalatrax commented on src/hints/math/testLessThanOrEqual.ts:
```diff
@@ -0,0 +1,62 @@
+ *
+ *
+ * @param {VirtualMachine} vm - The virtual machine instance.
+ * @param {ResOperand} lhs -  Pointer to the operand representing the left-hand side value in the comparison.
+ * @param {ResOperand} rhs -  Pointer to the operand representing the right-hand side value in the comparison.
```
Too verbose, "The left-hand side operand" is enough

zmalatrax reviewed (commented):
good, just some details on the JSDoc :)

zmalatrax commented on src/hints/math/testLessThanOrEqual.ts:
```diff
@@ -0,0 +1,59 @@
+/**
+ * TestLessThanOrEqual hint
+ *
+ * Store true at dst if value at lhs is inferior or equal to value at rhs, false otherwise.
```
Good, just add backticks for dst, lhs and rhs (`dst`, `lhs` and ` rhs`)

enitrat commented on cairo_programs/cairo/hints/test_less_than_or_equal_false.cairo:
```diff
@@ -0,0 +1,3 @@
+fn main() -> bool {
+    12_u32 <= 10_u32
+}
```
missing new lines at end of file:

in your vscode config add

```
  "files.insertFinalNewline": true,
  "files.trimFinalNewlines": true,
  "files.trimTrailingWhitespace": true,
  ```

zmalatrax replied:
remove this file

LazarusAA commented:
Forgot to mention that this closes #133

zmalatrax reviewed (approved):
Good job

zmalatrax commented on cairo_programs/cairo/hints/test_less_than_or_equal_true.cairo:
```diff
@@ -0,0 +1,5 @@
+fn main() -> bool {
+    let a = 10_u32;
+    let b = 12_u32;
+    a <= b && a <= a
+}
```
Rename this file to `span_slice.cairo`

The Cairo function that uses the `TestLessThanOrEqual` hint is the `array_slice`, when slicing a span.

How to find such functions to test a hint though ?
- [Search](https://github.com/search?q=repo%3Astarkware-libs%2Fcairo+%22hint+TestLessThanOrEqual%22&type=code) for a libfunc in the Cairo compiler that uses that hint: "hint TestLessThanOrEqual"
- Find one that is not using too many hints, or unimplemented ones, here [`build_array_slice`](https://github.com/starkware-libs/cairo/blob/86b6b559ea15359ffaf35b3aa6474836179c57ce/crates/cairo-lang-sierra-to-casm/src/invocations/array.rs#L311)
- [Search ](https://github.com/search?q=repo%3Astarkware-libs%2Fcairo+%22array_slice%22&type=code)for the corresponding function name in the Cairo corelib `array_slice()`
- Check if the function is part of the public API or find a function that uses it, here [`slice` for Span](https://github.com/starkware-libs/cairo/blob/86b6b559ea15359ffaf35b3aa6474836179c57ce/corelib/src/array.cairo#L363-L374)
- Make a Cairo program that uses this function

Usually, the Cairo corelib gives usage example to use this function.

A Cairo program like this can be enough:

```cairo
fn main() {
    let span = array![1, 2, 3].span();
    let _slice = span.slice(0, 1);
}
```

LazarusAA replied:
Done, thank you for the detailed explanation :)

zmalatrax reviewed (commented):
We've previously talked about the fact that Cairo 2.7.0 doesn't use the `TestLessThanOrEqual` hint when doing arithmetic comparisons. I said that we could let the Cairo programs as is, and use an older version of the compiler to test it, until we find a Cairo program that uses it for the latest Cairo version.

I've found a Cairo libfuncs that uses the `TestLessThanOrEqual` hint, so it can be tested with Cairo 2.7.0 :)

Note: After further thinking, if the latest Cairo version doesn't use a specific hint for one operation, another program should be written, using another operation that still uses this hint, not defaulting on "using a previous Cairo version".

zmalatrax commented on cairo_programs/cairo/hints/span_life.cairo:
```diff
@@ -0,0 +1,4 @@
+fn main() {
```
I don't understand the filename `span_life.cairo`
Maybe rename it to `span_slice.cairo` ?
